tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+            RoleType::Targets => {
+                if consistent_snapshot {
+                    format!("{}.targets.json", self.signed.signed.version())
+                } else {
+                    "targets.json".to_string()
+                }
+            }
+            RoleType::Snapshot => {
+                if consistent_snapshot {
+                    format!("{}.snapshot.json", self.signed.signed.version())
+                } else {
+                    "targets.json".to_string()
+                }
+            }
+            RoleType::Timestamp => "timestamp.json".to_string(),
+            RoleType::Root => format!("{}.root.json", self.signed.signed.version()),
+        };
+
+        let path = outdir.join(filename);
+        std::fs::write(&path, &self.buffer).context(error::FileWrite { path })
+    }
+}
+
+/// A set of signed TUF Repository metadata.
+///
+/// This metadata represents a signed TUF repository, but without the actual
+/// targets, the repository cannot be used. It is up to ther user to ensure
+/// all the target files are in the proper directories along with the
+/// contained metadata. There are convenience methods provided to help with
+/// this.
+#[derive(Debug)]
+pub struct SignedRepository {
+    root: SignedRole<Root>,
+    targets: SignedRole<Targets>,
+    snapshot: SignedRole<Snapshot>,
+    timestamp: SignedRole<Timestamp>,
+}
+
+impl SignedRepository {
+    /// Writes the metadata to the given directory. If consistent snapshots
+    /// are used, the appropriate files are prefixed with their `sha256`.
+    pub fn write<P>(&self, outdir: P) -> Result<()>
+    where
+        P: AsRef<Path>,
+    {
+        let consistent_snapshot = self.root.signed.signed.consistent_snapshot;
+        self.root.write(&outdir, consistent_snapshot)?;
+        self.targets.write(&outdir, consistent_snapshot)?;
+        self.snapshot.write(&outdir, consistent_snapshot)?;
+        self.timestamp.write(&outdir, consistent_snapshot)
+    }
+
+    /// Crawls a given directory and symlinks any targets found to the given
+    /// "out" directory.
+    ///
+    /// For each file found in the `indir`, the method gets the filename and
+    /// if the filename exists in `Targets`, the file's sha256 is compared
+    /// against the data in `Targets`. If this data does not match, the
+    /// method will fail. If all is well, the taret is symlinked.
+    pub fn link_targets<P1, P2>(&self, indir: P1, outdir: P2) -> Result<()>
+    where
+        P1: AsRef<Path>,
+        P2: AsRef<Path>,
+    {
+        std::fs::create_dir_all(&outdir).context(error::DirCreate {
+            path: &outdir.as_ref(),
+        })?;
+        // Get the absolute path of the indir and outdir
+        let abs_indir = std::fs::canonicalize(&indir).context(error::AbsolutePath {
+            path: indir.as_ref().to_owned(),
+        })?;
+        let abs_outdir = std::fs::canonicalize(&outdir).context(error::AbsolutePath {
+            path: outdir.as_ref().to_owned(),
```

Comment:
```suggestion
            path: indir.as_ref(),
        })?;
        let abs_outdir = std::fs::canonicalize(&outdir).context(error::AbsolutePath {
            path: outdir.as_ref(),
```
Snafu should handle cloning for you, and only in the error case; this would clone even in the success case.  (If you need to do it manually for some reason I'm missing, `.with_context(|| bla)` will only run in the error case, which is good if the error creation takes a few cycles...)